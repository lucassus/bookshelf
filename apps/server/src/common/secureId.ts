import { SecureId } from "@bookshelf/secure-id";

const secureId = new SecureId<string>({
  separator: "-"
});

export const toInternalIdAndType = (id: string) =>
  secureId.toInternalAndType(id);

export const toInternalId = (id: string): number =>
  Number(secureId.toInternal(id));

export const toExternalId = (entity: { id: number }): string =>
  secureId.toExternal(
    entity.id,
    Object.getPrototypeOf(entity).constructor.name
  );
