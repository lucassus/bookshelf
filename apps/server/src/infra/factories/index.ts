export { createAuthor } from "./createAuthor";
export { createAvatar } from "./createAvatar";
export { createBook } from "./createBook";
export { createBookCopy } from "./createBookCopy";
export { createUser } from "./createUser";
export { createReview } from "./createReview";
