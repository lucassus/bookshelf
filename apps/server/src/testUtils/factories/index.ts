export { createAuthor } from "./createAuthor";
export { createAvatar } from "./createAvatar";
export { createBook } from "./createBook";
export { createBookCopy } from "./createBookCopy";
export { createUser } from "./createUser";
