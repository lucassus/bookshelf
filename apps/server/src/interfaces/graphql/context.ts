import { PubSub } from "apollo-server-express";
import express from "express";
import { ConnectionContext, ExecutionParams } from "subscriptions-transport-ws";
import { Container } from "typedi";

import { ASSETS_BASE_URL } from "../../infra/config";
import { User } from "../../infra/database/entity";
import {
  buildAuthorsLoader,
  buildBooksLoader,
  buildUsersLoader
} from "../../infra/dataLoaders";
import { authenticateRequest } from "../../infra/support/authentication";

// eslint-disable-next-line import/order
import WebSocket = require("ws");

const pubsub = new PubSub();

export interface Context {
  req: express.Request;
  res: express.Response;
  container: typeof Container;
  assetsBaseUrl: string;
  authorsLoader: ReturnType<typeof buildAuthorsLoader>;
  booksLoader: ReturnType<typeof buildBooksLoader>;
  usersLoader: ReturnType<typeof buildUsersLoader>;
  pubsub: PubSub;
  currentUser?: User;
}

export interface AuthenticatedContext extends Context {
  currentUser: User;
}

export const onSubscriptionConnect = async (
  params: any,
  ws: WebSocket,
  context: ConnectionContext
) => {
  const currentUser = await authenticateRequest(context.request);
  return { currentUser };
};

export const createContext = async ({
  req,
  res,
  connection
}: {
  req: express.Request;
  res: express.Response;
  connection?: ExecutionParams;
}): Promise<Context | AuthenticatedContext> => {
  const currentUser = connection
    ? connection.context.currentUser
    : await authenticateRequest(req);

  return {
    assetsBaseUrl: ASSETS_BASE_URL,
    authorsLoader: buildAuthorsLoader(),
    booksLoader: buildBooksLoader(),
    usersLoader: buildUsersLoader(),
    container: Container,
    currentUser,
    pubsub,
    req,
    res
  };
};
