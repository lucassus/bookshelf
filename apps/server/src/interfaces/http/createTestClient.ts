import cookie from "cookie";
import express from "express";
import request from "supertest";

import { api as apiRoutes } from "./api/routes";
import { AUTH_COOKIE_NAME } from "~/infra/config";
import { User } from "~/infra/database/entity";
import { generateAuthToken } from "~/infra/support/authentication";

export function createTestClient({
  authToken,
  currentUser
}: { currentUser?: User; authToken?: string } = {}) {
  const app = express();

  app.use((req, res, next) => {
    const token = authToken || (currentUser && generateAuthToken(currentUser));

    if (token) {
      req.headers = {
        cookie: cookie.serialize(AUTH_COOKIE_NAME, token)
      };
    }

    next();
  });

  app.use("/api", apiRoutes);

  return request(app);
}
