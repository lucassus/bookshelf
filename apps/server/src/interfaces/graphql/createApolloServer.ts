import { ApolloServer } from "apollo-server-express";

import { Environment, ENVIRONMENT } from "../../infra/config";
import { createContext, onSubscriptionConnect } from "./context";
import { rootSchema } from "./rootSchema";

export const createApolloServer = () =>
  new ApolloServer({
    schema: rootSchema,
    context: createContext,
    subscriptions: {
      onConnect: onSubscriptionConnect
    },
    plugins: [
      {
        requestDidStart: (requestContext) => {
          console.log(
            `\n\n${requestContext.request.operationName} operation started`
          );
        }
      }
    ],
    debug: ENVIRONMENT === Environment.development,
    tracing: ENVIRONMENT === Environment.development,
    introspection: ENVIRONMENT !== Environment.production,
    playground: {
      settings: {
        "request.credentials": "include"
      }
    },
    engine:
      ENVIRONMENT === Environment.production
        ? {
            reportSchema: true,
            debugPrintReports: true
          }
        : false
  });
