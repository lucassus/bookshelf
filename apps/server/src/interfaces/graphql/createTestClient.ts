import { ApolloServer } from "apollo-server-express";
import {
  ApolloServerTestClient,
  createTestClient as createApolloTestClient
} from "apollo-server-testing";
import cookie from "cookie";
import express from "express";
import httpMocks from "node-mocks-http";

import { AUTH_COOKIE_NAME } from "../../infra/config";
import { User } from "../../infra/database/entity";
import { generateAuthToken } from "../../infra/support/authentication";
import { createContext } from "./context";
import { rootSchema } from "./rootSchema";

export function createTestClient({
  currentUser,
  res = httpMocks.createResponse()
}: {
  currentUser?: User;
  res?: express.Response;
} = {}): ApolloServerTestClient {
  const config = {
    schema: rootSchema,
    context: () => {
      const req = httpMocks.createRequest();

      if (currentUser) {
        req.headers = {
          cookie: cookie.serialize(
            AUTH_COOKIE_NAME,
            generateAuthToken(currentUser)
          )
        };
      }

      return createContext({ req, res });
    }
  };

  return createApolloTestClient(new ApolloServer(config));
}
