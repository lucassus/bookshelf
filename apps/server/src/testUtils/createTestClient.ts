import { ApolloServer } from "apollo-server-express";
import {
  ApolloServerTestClient,
  createTestClient as createApolloTestClient
} from "apollo-server-testing";
import express from "express";
import httpMocks from "node-mocks-http";

import { generateAuthToken } from "../common/authentication";
import { AUTH_COOKIE_NAME } from "../config";
import { User } from "../database/entity";
import { buildContext } from "../graphql/context";
import { rootSchema } from "../graphql/rootSchema";

export function createTestClient({
  currentUser,
  res = httpMocks.createResponse()
}: {
  currentUser?: User;
  res?: express.Response;
} = {}): ApolloServerTestClient {
  const config = {
    schema: rootSchema,
    context: () => {
      const req = httpMocks.createRequest();

      if (currentUser) {
        req.cookies = {
          [AUTH_COOKIE_NAME]: generateAuthToken(currentUser)
        };
      }

      return buildContext({ req, res });
    }
  };

  return createApolloTestClient(new ApolloServer(config));
}
