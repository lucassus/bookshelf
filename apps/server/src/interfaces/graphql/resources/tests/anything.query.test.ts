import { gql } from "apollo-server-express";

import { createAuthor, createBook, createUser } from "~/infra/factories";
import { toExternalId } from "~/infra/support/secureId";
import { createTestClient } from "~/interfaces/graphql/createTestClient";

describe("anything query", () => {
  const GetAnythingQuery = gql`
    query($id: ID!) {
      anything(id: $id) {
        __typename

        ... on Author {
          id
          name
          books {
            ...BookFragment
          }
        }

        ...BookFragment

        ...UserFragment
      }
    }

    fragment BookFragment on Book {
      id
      title
      description
    }

    fragment UserFragment on User {
      id
      name

      ... on ProtectedUser {
        email
        isAdmin
      }
    }
  `;

  it("fetches Author", async () => {
    // Given
    const author = await createAuthor();
    await createBook({ author });
    await createBook({ author });
    await createBook({ author });

    // When
    const res = await createTestClient().query({
      query: GetAnythingQuery,
      variables: { id: toExternalId(author) }
    });

    // Then
    expect(res.errors).toBe(undefined);
    expect(res.data).toMatchSnapshot();
  });

  it("fetches Book", async () => {
    // Given
    const book = await createBook();

    // When
    const res = await createTestClient().query({
      query: GetAnythingQuery,
      variables: { id: toExternalId(book) }
    });

    // Then
    expect(res.errors).toBe(undefined);
    expect(res.data).toMatchSnapshot();
  });

  it("fetches User", async () => {
    // Given
    const user = await createUser();

    // When
    const res = await createTestClient({ currentUser: user }).query({
      query: GetAnythingQuery,
      variables: { id: toExternalId(user) }
    });

    // Then
    expect(res.errors).toBe(undefined);
    expect(res.data).toMatchObject({
      anything: {
        __typename: "ProtectedUser",
        id: expect.any(String),
        name: "Randolph Satterfield",
        email: "Lauryn.Crist@yahoo.com",
        isAdmin: false
      }
    });
  });
});

it("fetches with aliases", async () => {
  // Given
  const book = await createBook();
  const author = await createAuthor();
  const user = await createUser();

  // When
  const res = await createTestClient().query({
    query: gql`
      query(
        $bookId: ExternalID!
        $authorId: ExternalID!
        $userId: ExternalID!
      ) {
        something: book(id: $bookId) {
          ... on Book {
            id
            externalId: id
            headline: title
            description
          }
        }

        author(id: $authorId) {
          ... on Author {
            id
            headline: name
            description: bio
          }
        }

        user(id: $userId) {
          ... on User {
            id
            headline: name
            description: info
          }
        }
      }
    `,
    variables: {
      bookId: toExternalId(book),
      authorId: toExternalId(author),
      userId: toExternalId(user)
    }
  });

  // Then
  expect(res.errors).toBe(undefined);
  expect(res.data).toMatchSnapshot();
});
