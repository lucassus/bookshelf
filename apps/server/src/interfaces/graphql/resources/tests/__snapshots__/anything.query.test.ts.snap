// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`anything query fetches Author 1`] = `
{
  "anything": {
    "__typename": "Author",
    "books": [
      {
        "description": "Natus aliquam iste rem.",
        "id": "Mi1Cb29r",
        "title": "Dolorem Alias Qui Asperiores",
      },
      {
        "description": "Ipsa reiciendis nam.",
        "id": "MS1Cb29r",
        "title": "Fugit Facilis Mollitia",
      },
      {
        "description": "Iste aliquid odit aperiam dolore voluptatibus nemo.",
        "id": "My1Cb29r",
        "title": "Sit Dolore",
      },
    ],
    "id": "MS1BdXRob3I=",
    "name": "Miss Sadie Deckow-Reynolds",
  },
}
`;

exports[`anything query fetches Book 1`] = `
{
  "anything": {
    "__typename": "Book",
    "description": "Ipsa reiciendis nam.",
    "id": "MS1Cb29r",
    "title": "Fugit Facilis Mollitia",
  },
}
`;

exports[`fetches with aliases 1`] = `
{
  "author": {
    "description": "Iste rem consequatur eum sit.",
    "headline": "Mr. Leonard Deckow V",
    "id": "Mi1BdXRob3I=",
  },
  "something": {
    "description": "Ipsa reiciendis nam.",
    "externalId": "MS1Cb29r",
    "headline": "Fugit Facilis Mollitia",
    "id": "MS1Cb29r",
  },
  "user": {
    "description": "Laudantium maiores provident quam aperiam nulla.",
    "headline": "Christina Greenholt III",
    "id": "MS1Vc2Vy",
  },
}
`;
